types/mongoose/test/model.ts

```diff
@@ -476,3 +476,34 @@ LocModel.updateMany({ name: 'foo' }, { name: 123 });
 LocModel.updateMany({ name: 'foo' }, { $pull: { facilities: 123 } });
 // $ExpectError
 LocModel.updateMany({ name: 'foo' }, { $push: { coords: 'bar' } });
+
+
+// $ExpectError
+LocModel.create({ address: "foo", coords: [1, 2], facilities: ["foo", "bar"] });
+
+LocModel.create({ address: "foo", coords: [1, 2], facilities: ["foo", "bar"], name: "bar", openingTimes: ["foo"], rating: 10, reviews: ["foo"] });
+LocModel.create<{address: string}>({ address: "foo" });
+
+interface ModelWithFunction extends mongoose.Document {
+    name: string;
+
+    someFunc: () => any;
+
+    deeperFuncTest: {
+        test: string;
+        // ensure this doesn't turn required
+        optionalTest?: string;
+
+        someFunc: () => any;
+    }
+}
+const modelWithFunctionSchema = new mongoose.Schema({
+    name: { type: String, required: true },
+});
+
+var ModelWithFunctionInSchema = mongoose.model<ModelWithFunction>("ModelWithFunction", modelWithFunctionSchema);
```

Comment:
To workaround the functions inside this type and remove the necessity of the `DeepNonFunctionProperties` wrapper, the following pattern should be used:

```ts
interface FooDocument extends mongoose.Document {
    name: string;
    deeperFuncTest: {
        test: string;
        // ensure this doesn't turn required
        optionalTest?: string;
    };
}
interface FooModel extends mongoose.Model<FooDocument> {
    someFunc: () => any;
    deeperFuncTest: {
        someFunc: () => any;
    };
}

const fooSchema = new mongoose.Schema({
    name: { type: String, required: true },
});

var Foo = mongoose.model<FooModel, FooDocument>("Foo", fooSchema);
```

Reply:
See my other comment. Let me know what you think.

Reply:
For more context on why there could be functions inside models, consider [typegoose](https://github.com/typegoose/typegoose#motivation).

It makes it easy to generate schemas based on decorated classes. Since these classes can potentially also include methods in them, a way to omit them for creation is required.

Otherwise, it leads to additional boilerplate code, where a separate class/interface needs to be defined just to get `.create()` to comply.

Reply:
Additionally, in practice if you have something like

```ts
class Foo {
  bar: Map<string, Bar>;
  privateProp: string;

  setBar(barLike: any) {
    this.bar = convertSomehow(someBar);
  }
}

class Bar {
  title: string;
  
  someFunction(someParam: any) {
    
  }
}
```

Now you need interfaces that exclude functions for both `Foo` and `Bar` and you need to connect `FooSchema.bar` to the `BarSchema` type. 

The more complex and intertwined the models, the harder this becomes to do, up to a point where it becomes unmaintainable.

I feel the right thing to do is just omit functions by default in the type. When using plain mongoose they don't matter anyway.

Reply:
I'm not sure I am comfortable with this change, so I'll leave other reviewers take a stab at this.

Reply:
FWIW, it's not really a "change", but an addition. :) Nothing about current behavior really changes.

Reply:
I know :)

Could you perhaps use these utility types from `mongodb` instead of your `DeepNonFunctionProperties` utility type?

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mongodb/index.d.ts#L1239-L1246